packages/core/content-manager/server/src/controllers/relations.ts

```diff
@@ -201,106 +188,104 @@ export default {
       });
     }
 
-    if (_q) {
-      // searching should be allowed only on mainField for permission reasons
-      const _filter = isOperatorOfType('where', query._filter) ? query._filter : '$containsi';
-      addFiltersClause(queryParams, { [mainField]: { [_filter]: _q } });
-    }
-
-    if (currentEntityId) {
-      // If we have been given an entity id, we need to filter out the
-      // relations that are already linked to the current entity
+    if (id) {
+      // If we have been given an id (document or entity), we need to filter out the
+      // relations that are already linked to the current id
       const subQuery = strapi.db.queryBuilder(sourceUid);
 
-      // The alias refers to the DB table of the target model
+      // The alias refers to the DB table of the target content type model
       const alias = subQuery.getAlias();
 
       const where: Record<string, any> = {
-        id: currentEntityId,
         [`${alias}.id`]: { $notNull: true },
-        // Always find available draft entries, as we will be potentially
-        // connecting them from the CM edit view
-        [`${alias}.published_at`]: { $notNull: isPublished },
+        [`${alias}.document_id`]: { $notNull: true },
       };
 
+      const isSourceComponent = sourceModelType === 'component';
+      if (isSourceComponent) {
+        // If the source is a component, we need to filter by the component's
+        // numeric entity id
+        where.id = id;
+      } else {
+        // If the source is a content type, we need to filter by document id
+        where.document_id = id;
+      }
+
       if ((idsToInclude?.length ?? 0) !== 0) {
         where[`${alias}.document_id`].$notIn = idsToInclude;
       }
 
       const knexSubQuery = subQuery
         .where(where)
         .join({ alias, targetField })
-        .select(`${alias}.id`)
+        .select(`${alias}.document_id`)
```

Comment:
should we add the locale and publication state filter here too?

Reply:
Do you think this is okay?
[ddf8133](https://github.com/strapi/strapi/pull/19681/commits/ddf813331f393c9e53e689baa5599443009081d9)

Reply:
looks good! :)